### ISSUE TITLE
mcscf.addons.state_average_ and state_average_mix_ inheritance question

### ISSUE THREAD

**Issue by MatthewRHermes** [pyscf/pyscf#305]
Is there a reason that the state_average_ and state_average_mix_ replace the "\_finalize" method attribute with an instance attribute (lines 678-696 and 879-897), instead of making a child class as was done just a few lines above in "FakeCISolver"?  I ask because the current code makes combining different child classes awkward.  For instance, if I try to make a scanner out of the state_average_ method:

```
from pyscf import gto, scf, mcscf
from pyscf.lib import logger

mol = gto.M(atom='N 0 0 0; N 0 0 1.2', basis='6-31g', output='n2_example.log', verbose=logger.INFO)
mf = scf.RHF(mol)
mf.kernel ()
mc = mcscf.CASSCF(mf, 6, 6)
mc.state_average_ ()
mc_scanner = mc.as_scanner ()
mc_scanner ('N 0 0 0; N 0 0 1.1')
```
This happens:
```
Traceback (most recent call last):
  File "minimal_example.py", line 10, in <module>
    mc_scanner ('N 0 0 0; N 0 0 1.1')
  File "/home/gagliard/herme068/lib/python3.4/site-packages/pyscf/mcscf/mc1step.py", line 530, in __call__
    e_tot = self.kernel(mo, self.ci)[0]
  File "/home/gagliard/herme068/lib/python3.4/site-packages/pyscf/mcscf/mc1step.py", line 798, in kernel
    self._finalize()
  File "/home/gagliard/herme068/lib/python3.4/site-packages/pyscf/mcscf/addons.py", line 680, in _finalize
    old_finalize()
  File "/home/gagliard/herme068/lib/python3.4/site-packages/pyscf/mcscf/casci.py", line 835, in _finalize
    for i, e in enumerate(self.e_cas):
TypeError: 'NoneType' object is not iterable
```
Which is clearly because the "self" in the final line is not the same object as the "self" in the second and third lines, because I made a child class _after_ calling state\_average\_, and the "old\_finalize" call in the fourth line refers to a method of the instance that was _before_ calling state\_average\_.  Of course, if I just call state\_average\_ on the 'mc_scanner' object instead of the 'mc' object, everything seems to work fine.  But it's harder to figure out how to do that in, for instance, geometry optimizations.

I guess I was tempted to just change it to a child class with a new "_finalize" method and make a pull request but I wanted to ask if there was a reason for doing it this way before I did that.

**Comment by sunqm** [comment]
@MatthewRHermes  Thanks for catching the problem. When implementing the state_average function, I didn't consider the geometry optimization. It was a dirty fix to support the state-average functionality, and to patch the mcscf object inplace. A child class is a better choice in this case.

Note the state_average_ function modify the mcscf object inplace. When introducing the child class, the instance of child class may not be able to modify the input mcscf object. It's different to the behavior of state_average_ function. So please put the child class in a new function, e.g., state_average (without _). I think there is a workaround, though a bit hacky, to make the state_average_ function backward compatible. The state_average_ function first executes the child-class-version state-average function and get the child class. Then the original mc.__class__ can be replaced by the generated child class " mc.__class__ = state_average(mc).__class__".  I didn't realize other treatments at this moment. You can choose other solutions as you like.

**Comment by sunqm** [comment]
Fixed in PR #306
